### PR TITLE
New resource: aws_autoscaling_instance_refresh

### LINKED ISSUE #13785: Support ASG Instance Refresh
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When an auto-scaling group is updated in Terraform (e.g. modification of the Launch Template version), its instances remain unchanged, and a complete roll-out requires intervention by other means.

This is in contrast to [AWS::AutoScaling::AutoScalingGroup](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html) in CloudFormation, which orchestrates a roll-out, and waits until it's completed or times out.

AWS have recently introduced [ASG Instance Refresh](https://docs.aws.amazon.com/autoscaling/ec2/userguide/asg-instance-refresh.html), which can be used to recycle the instances of an auto-scaling group.

Use ASG Instance Refresh to provide the user the option to automatically refresh an entire ASG in response to changes to an auto-scaling group. The proposed change introduces a new `aws_autoscaling_group` block, `instance_refresh`, that instructs Terraform to create and monitor an instance refresh in response to any changes to the ASG's properties, except for (1) those that do not affect individual instances, such as `max_size`; and (2) properties explicitly ignored by way of lifecycle.

The `instance_refresh` would be disabled by default, thus ensuring backwards compatibility.

### New or Affected Resource(s)

* aws_autoscaling_group

### Potential Terraform Configuration

```hcl
resource "aws_launch_template" "foobar" {
  name_prefix   = "foobar"
  image_id      = "ami-1a2b3c"
  instance_type = "t2.micro"
}

resource "aws_autoscaling_group" "bar" {
  availability_zones = ["us-east-1a"]
  desired_capacity   = 1
  max_size           = 1
  min_size           = 1

  launch_template {
    id      = "${aws_launch_template.foobar.id}"
    version = "${aws_launch_template.foobar.latest_version}"
  }

  instance_refresh {
    enabled = true
    instance_warmup = 300
    min_healthy_percentage = 90
  }
}
```

### References

* https://docs.aws.amazon.com/autoscaling/ec2/userguide/asg-instance-refresh.html
* https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#AutoScaling.DescribeInstanceRefreshes
* https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#AutoScaling.StartInstanceRefresh
* https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#AutoScaling.CancelInstanceRefresh

### PR DESCRIPTION
**Pull request by roberth-k**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #13785

Supersedes #13791. Representing Instance Refresh as a separate resource enables
consistent failure behaviour when waiting for refresh completion.

The `aws_autoscaling_instance_refresh` block includes a `triggers` block, which
allows additional flexibility. The default recommendation is to use the new
`instance_refresh_token` token attribute of the auto-scaling group, which changes
every time the ASG is modified in a way that would warrant a refresh (e.g. new
launch template or VPC configuration).

Example:

```hcl
resource "aws_launch_configuration" "test" {
  image_id      = "ami-123456789"
  instance_type = "t2.micro"
}

resource "aws_autoscaling_group" "test" {
  availability_zones   = ["us-west-2"]
  min_size             = 1
  max_size             = 2
  launch_configuration = aws_launch_configuration.test.name
}

resource "aws_autoscaling_instance_refresh" "test" {
  autoscaling_group_name  = aws_autoscaling_group.test.name
  min_healthy_percentage  = 50
  instance_warmup_seconds = 5
  strategy                = "Rolling"

  triggers = {
    token = aws_autoscaling_group.test.instance_refresh_token
  }
}
```

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
FEATURES:
- **New Resource:** `aws_autoscaling_instance_refresh`

ENHANCEMENTS:
- resource/aws_autoscaling_group: Add `instance_refresh_token` attribute
```

Output from acceptance testing:

```
$ clear && make testacc TESTARGS='-run=TestAccAWSAutoscalingInstanceRefresh_'
==> Checking that code complies with gofmt requirements...
--- PASS: TestAccAWSAutoscalingInstanceRefresh_alreadyOngoing (143.95s)
--- PASS: TestAccAWSAutoscalingInstanceRefresh_cancelOnTimeout (160.09s)
--- PASS: TestAccAWSAutoscalingInstanceRefresh_basic (346.45s)
--- PASS: TestAccAWSAutoscalingInstanceRefresh_disappears (348.04s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       349.391s

$ make testacc TESTARGS='-run=TestAccAWSAutoScalingGroup_'
==> Checking that code complies with gofmt requirements...
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_Override_WeightedCapacity (146.81s)
--- PASS: TestAccAWSAutoScalingGroup_basic (231.18s)
--- PASS: TestAccAWSAutoScalingGroup_InstanceRefresh_Triggers (267.39s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_Override_InstanceType (43.81s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_LaunchTemplateSpecification_Version (44.19s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_LaunchTemplateSpecification_LaunchTemplateName (42.14s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotMaxPrice (45.88s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotInstancePools (44.02s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotAllocationStrategy (42.42s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandPercentageAboveBaseCapacity (44.06s)
--- PASS: TestAccAWSAutoScalingGroup_ALB_TargetGroups_ELBCapacity (290.56s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_UpdateToZeroOnDemandBaseCapacity (43.24s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandBaseCapacity (41.67s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandAllocationStrategy (41.97s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy (42.13s)
--- PASS: TestAccAWSAutoScalingGroup_LaunchTemplate_IAMInstanceProfile (56.35s)
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate_update (82.41s)
--- PASS: TestAccAWSAutoScalingGroup_classicVpcZoneIdentifier (55.51s)
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate (44.47s)
--- PASS: TestAccAWSAutoScalingGroup_launchTempPartitionNum (44.54s)
--- PASS: TestAccAWSAutoScalingGroup_enablingMetrics (140.59s)
--- PASS: TestAccAWSAutoScalingGroup_initialLifecycleHook (187.04s)
--- PASS: TestAccAWSAutoScalingGroup_TargetGroupArns (180.74s)
--- PASS: TestAccAWSAutoScalingGroup_ALB_TargetGroups (112.25s)
--- PASS: TestAccAWSAutoScalingGroup_MaxInstanceLifetime (44.74s)
--- PASS: TestAccAWSAutoScalingGroup_serviceLinkedRoleARN (50.89s)
--- PASS: TestAccAWSAutoScalingGroup_withMetrics (43.21s)
--- PASS: TestAccAWSAutoScalingGroup_VpcUpdates (48.13s)
--- PASS: TestAccAWSAutoScalingGroup_LoadBalancers (663.40s)
--- PASS: TestAccAWSAutoScalingGroup_suspendingProcesses (170.76s)
--- PASS: TestAccAWSAutoScalingGroup_withPlacementGroup (129.83s)
--- PASS: TestAccAWSAutoScalingGroup_terminationPolicies (70.89s)
--- PASS: TestAccAWSAutoScalingGroup_tags (199.31s)
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer (303.30s)
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer_ToTargetGroup (363.79s)
--- PASS: TestAccAWSAutoScalingGroup_autoGeneratedName (41.13s)
--- PASS: TestAccAWSAutoScalingGroup_namePrefix (51.76s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       1500.890s
```

### CODE REVIEW

**Comment by jamroks** [comment]
Hi, would be great to have this merge , since I don't see anything blocking it technically speaking , maybe I'm missing something . 

thanks for the work on this @roberth-k

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by sanspace** [comment]
Would this one ever get merged!? 😨

**Comment by brewkode** [comment]
Does someone know when this will be merged ?

File: website/docs/r/autoscaling_instance_refresh.html.markdown [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+~> **NOTE:** The time it takes to refresh an auto-scaling group is highly sensitive to factors such
+as capacity and instance warmup time. The default timeout is
+intentionally a very short duration. Ensure that you have chosen a create timeout
+that appropriately estimates the expected duration of a successful rollout.
+
+## Import
+
+`aws_autoscaling_instance_refresh` can be imported using a concatenation of the
+`autoscaling_group_name` and `instance_refresh_id` by an underscore (`_`). For example:
```
Comment (Vivalldi) [inline website/docs/r/autoscaling_instance_refresh.html.markdown:99]: The separator is actually a forward-slash (`/`) correct?

Comment (roberth-k) [reply website/docs/r/autoscaling_instance_refresh.html.markdown]: Correct -- thanks!

**Comment by gdavison** [comment]
Hi @roberth-k, thanks for the two PRs for this feature. We've been putting a lot of thought into the best way to address instance refresh. I'll follow up with you later this week

**Comment by ishworg** [comment]
I feel like a lot of folks out there (including *myself*) are waiting on this pull request to land on a release for a while now.
Here's my two cent how I currently do it using `awscli`:

```bash

docker run --rm \
    -e AWS_ACCESS_KEY_ID \
    -e AWS_SECRET_ACCESS_KEY \
    -e AWS_REGION \
    awscli:latest \
      aws --region ${AWS_REGION} \
        autoscaling start-instance-refresh \
        --auto-scaling-group-name "my-${bg_active_stack}" \
        --preferences '{"InstanceWarmup": 600, "MinHealthyPercentage": 90}'
```

Hope it helps.

**Comment by smiller171** [comment]
@gdavison Did that follow-up happen?

Currently we're using CloudFormation (through Terraform) to handle ASGs _specifically_ for these kinds of rolling updates

**Comment by tsalinguepumpkin** [comment]
@gdavison any update on the matter ?

**Comment by nelg** [comment]
Tested this (compiled and tested locally), works OK for me.

It would be good to get this merged

**Comment by nashvan** [comment]
any progress here? is it going to be merged anytime soon?

**Pull request closed without merging**